highfidelity/hifi
# Entity list - dynamic columns

- Change entity list columns to be data-driven
- Add ability for columns to be resized using the borders on the header cells
- Add multiselect columns dropdown in place of the previous "show extra info" toggle to allow for columns to be added and removed

Addresses:
https://highfidelity.manuscript.com/f/cases/1318/

Test Plan:

For the QA retest, you only need to quickly verify that resizing columns, adding/removing columns, and sorting columns continues to work as expected.

- Enter Interface in desktop mode into any domain with many different types of entities
- Open Create app to the entity list
- Adjust your search radius or your camera view (plus hitting refresh on the list) if you don't have a scrollable list with 50+ entities or so
- Verify that at the top of the scrollbar in place of the old right arrow to toggle showing extra columns there is now a down arrow dropdown that opens to a list of selectable columns as shown below
![entitylistcolumndropdown](https://user-images.githubusercontent.com/31390055/48221048-d0323500-e345-11e8-93d3-1b6fba60e05d.png)
- Verify with the Columns dropdown open that when clicking anywhere outside of it (including when clicking to open the Types dropdown) that the Columns dropdown closes
- Verify the Types dropdown still performs as expected for adding/removing entity types
- Verify when opening the Types dropdown and that when clicking anywhere else outside of it (including when clicking to open the Columns dropdown) that the Types dropdown closes
- Verify when toggling the column checkboxes that the column is added or removed from the list as expected and that the other column widths adjust appropriately (they lose/gain width proportionally for columns added/removed based on their current percentage width of the table)
- Verify you can toggle sorting ascending/descending per column as expected and that the sort arrow shows at the top
- Verify when scrolling the list up and down that the column header row stays fixed at the top
- Verify when hovering over any of the borders between column headers that a resize cursor indicator appears
- Verify when selecting a column header border to resize that dragging it left lowers the left columns width by taking the width from the right column, and vice-versa, and that each row's cell for that column resizes with the header in parallel maintaining it's cell contents within the column header boundaries
- Verify you cannot drag the columns to width below the minimum which is as shown below
![entitylistcolumnwidth](https://user-images.githubusercontent.com/31390055/48223377-cdd2d980-e34b-11e8-947b-70e5bb4944ea.png)
- Verify when resizing the entity list window that everything maintains their size proportions in the newly adjusted window and that it appears the same
- Verify when double clicking a cell in the name column to rename the entity that the black rename text field that appears is sized appropriately within the resized name cell and performs as expected
- Verify when lowering the radius or applying a search filter to reduce the entity list to not be scrollable that the columns adjust to fill the scrollbar space, and vice-versa
- Stress test adding/removing different columns, resizing the columns, sorting the columns, and resizing the window
